url_matcher/patterns.py

```diff
@@ -0,0 +1,265 @@
+        self.netloc_re: Optional[Pattern] = None
+        self.path_re: Optional[Pattern] = None
+        self.fragment_re: Optional[Pattern] = None
+        self.query_re_dict: Optional[Dict[str, Pattern]] = None
+        self._build_regexes()
+
+    def _build_regexes(self):
+        """
+        Builds the compiled regexes that can be used to match the pattern.
+        """
+        pscheme, pnetloc, ppath, pquery, pfragment = self.parsed
+        if pnetloc:
+            netloc_re = re.escape(pnetloc)
+            if not any((ppath, pquery, pfragment)):
+                # Also match subdomains if there is no path, query or fragment in the pattern
+                netloc_re = rf"(?:.*\.)?{netloc_re}"
+            netloc_re = f"^(?:www.)?{netloc_re}$"
+            self.netloc_re = re.compile(netloc_re, re.IGNORECASE)
+        if ppath:
+            self.path_re = _path_or_fragment_re(ppath)
+        if pfragment:
+            self.fragment_re = _path_or_fragment_re(pfragment)
+        if pquery:
+            pkvs = parse_qs(pquery, keep_blank_values=True)
+            query_re_dict = {}
+            for pparam, values in pkvs.items():
+                pparam = pparam.lower()
+                if "*" in pparam:
+                    warnings.warn(
+                        f"Wildcard expansion is only allowed for the values in the query parameter. Pattern: '{self.pattern}'",
+                        SyntaxWarning,
+                    )
+                    pparam = pparam.replace("*", "")
+                if not pparam:
+                    continue
+                param_re = fr"^(?:{'|'.join([_wildcard_re_escape(value) for value in values])})$"
+                query_re_dict[pparam] = re.compile(param_re, re.IGNORECASE)
+            self.query_re_dict = query_re_dict or None
+
+    def match(self, url: str) -> bool:
+        """
+        Return True if the url matches the pattern.
+        """
+        parsed = _urlparse(url)
+        if self.parsed.scheme:
+            if parsed.scheme != self.parsed.scheme:
+                return False
+        if self.netloc_re:
+            if not self.netloc_re.match(parsed.netloc):
+                return False
+        if self.path_re:
+            if not self.path_re.match(parsed.path):
+                return False
+        if self.fragment_re:
+            if not self.fragment_re.match(parsed.fragment):
+                return False
+        if self.query_re_dict:
+            kvs = parse_qs(parsed.query, keep_blank_values=True)
+            kvs = {k.lower(): v for k, v in kvs.items()}
+            # All params must be present in the URL
+            for param, param_re in self.query_re_dict.items():
+                if param not in kvs:
+                    return False
+                if not any(param_re.match(value) for value in (kvs[param])):
+                    return False
+        return True
+
+
+def _path_or_fragment_re(path_or_fragment: str) -> Pattern:
```

Comment:
_**optional:**_ Might as well be a static method of the `PatternMatcher` class.

Reply:
Addressed